### PR TITLE
Restrict to use non-comparable types in primary key

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Backward Incompatible Change

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Restrict to use of non-comparable data types (like `AggregateFunction`)  in keys (Sorting key, Primary key, Partition key, and so on).

### CODE REVIEW

**Comment by alesapin** [comment]
Related to https://github.com/ClickHouse/ClickHouse/issues/16548

File: src/Core/Field.h [hunk 1]
```diff
@@ -767,9 +767,10 @@ T & Field::get()
 #ifndef NDEBUG
     // Disregard signedness when converting between int64 types.
     constexpr Field::Types::Which target = TypeToEnum<NearestFieldType<ValueType>>::value;
-    assert(target == which
-           || (isInt64FieldType(target) && isInt64FieldType(which))
-           || target == Field::Types::Decimal64 /* DateTime64 fields */);
+    if (target != which
```
Comment (akuzm) [inline src/Core/Field.h:770]: This is probably going to slow down some things like `sumMap`.

Btw, just noticed that we now have Decimal64 here, but the author haven't added any justification as to why this check somehow doesn't apply to them :sigh:

**Review by akuzm — APPROVED** [review]

File: src/Core/Field.h (hunk 1 above)
Comment (alesapin) [reply src/Core/Field.h:770]: I think we should care about the clarity of errors in debug mode, rather than performance?)

Comment (akuzm) [reply src/Core/Field.h:770]: Somehow I didn't notice the NDEBUG around it, in debug mode it's OK.

**Comment by akuzm** [comment]
Fuzzer failure is this one https://github.com/ClickHouse/ClickHouse/issues/15889
Yandes sync broken in master.
01493_alter_remove_properties_zookeeper is fixed https://github.com/ClickHouse/ClickHouse/commit/ebd065c7e02c15e1202dc0ba8c0db8a2395a1c4a

**Comment by akuzm** [comment]
`00840_long_concurrent_select_and_drop_deadlock` is in CI fest https://github.com/ClickHouse/ClickHouse/issues/14414#issuecomment-719728663

**Pull request merged**